Add external DISM generations support. Fixes #5 & #25.

Shidell opened the pull request:
Adds a "DismUtilities" class that adds support for external DISM libraries across DISM generations, which allows the user to (optionally) load external DISM libraries instead of defaulting to look for DISM on the local system.

Shidell commented:
@josemesona, I removed my previous fork and started over from scratch in an effort to clean up my branch (and all the changes therein), and better familiarize myself not only with Git, but also GitHub.

Now that I'm a bit more familiar, I've created a new fork to "start clean", so to speak, and re-added the Utilities class we've talked about in a separate branch. 

Hopefully this is cleaner and easier to work with. I know you were interested in modifying the name (from "Utilities" to "DismUtilities"), and potentially integrating some of these changes into the Dism.Initialize() method. I've made the "DismUtilities" change, but haven't done anything else, as I also know this wrapper preserves the original native API mapping, and doesn't expand upon it.

I do like the fact that this managed API only wraps the native one, and leaves expanded functionality aside. Perhaps expanded functionality, such as this DISM generational library support, should remain in a "Utilities" (or other similar) classes, to preserve the clean mapping of the native API?

Ultimately, this decision is up to you. Feel free to modify this branch as you see fit to best integrate with this library. If you'd prefer I make the changes, just let me know how you'd like to see it implemented, and I can make another branch for you to review.

Once again, thanks for your patience while I'm getting up to speed with Git & GitHub.

josemesona commented:
I really like this functionality but again I'm just concerned that it moves away from being a pure wrapper.

If you make `DismUtilities` `internal` and add a new method to DismApi, I will gladly merge this.

```C#
public static partial class DismApi
{
    public static void InitializeEx(
        DismLogLevel logLevel,
        string logFilePath,
        string scratchDirectory,
        DismGeneration generation)
    {
      // Call DismUtilities.LoadDismGenerationLibrary(), throw an exception if it returns false
      // Call Initialize(logLevel, logFIlePath, scratchDirectory)
    }
}
```
Be sure to unload the library in `Shutdown` if the user specified calls `InitializeEx`.  You can have all of the overloads if you want so that you can just call `DismApi.Initailize(DismGeneration.Latest)`.  You'll have to move `DismGeneration` out of `DismUtilities` and make it public which is fine.

This leaves `DismUtilities` in there and exposes its basic functionality as an input to an "Ex" method which is common in C++.  What do you think?

Shidell commented:
I think that sounds like a good plan. I like the idea of using Ex methods, I just need to implement them.

I'll work to have something implemented shortly and update.

Shidell commented:
I modified the branch as we discussed. A couple quick points:

- I've added some "InitializeEx()" methods as we discussed which default to loading the latest DISM Generation library, assuming that's probably the best default setting. There is an overload where one can specify the library to load, but that introduces a problem: Now that the DismUtilities class is internal, a user cannot access the Properties in DismUtilities to quickly determine which DISM Generation libraries are available, if any.

- I added a Property to the main DismApi class to hold the DISM Generation loaded (if one was loaded) to work two ways: First, a user can retrieve it to determine what version of the DISM Generation library they have loaded (if one is loaded) and two, if it is not set to "DismGeneration.NotFound", then this modification knows to unload the DISM Generation library in Shutdown().

I wanted to make these changes for further discussion before being prepared for merging; thus, I need to modify some comments (specifically in DismUtilities) to be more pertinent depending on what implementation we settle on.

I'm not thrilled by having a Property in DismApi to represent the DISM Generation that's loaded, because it feels like we're polluting the "pure" nature of your wrapper cleanly mapping the Win32 DISM API. On the other hand, I need to store whether or not a DISM Generation library was loaded, and this implementation also lets the user check what's in use, which (I think) is useful.

What are you thoughts, now with a more complete version of what we had discussed?

josemesona commented on src/Microsoft.Dism/DismApi.cs:
```diff
@@ -11,7 +11,29 @@ namespace Microsoft.Dism
+        /// <summary>
+        /// Represents the DISM Generational Library initialized for use with the DismApi Wrapper (via InitializeEx()). Returns the specific DismGeneration in use; otherwise, returns DismGeneration.NotFound.
+        /// </summary>
+        public static DismGeneration DismGenerationInitialized
```
Please make this internal, an autoproperty, and set a initial value so you can get rid of the `dismGenerationInitialized` field and setting it in `Initialize`

```C#
internal static DismGeneration DismGenerationInitialized { get; set; } = DismGeneration.None;
```

josemesona commented on src/Microsoft.Dism/DismApi.cs:
```diff
@@ -834,7 +860,85 @@ public static void Initialize(DismLogLevel logLevel, string logFilePath)
+        /// <exception cref="DismException">When a failure occurs.</exception>
+        public static void InitializeEx(DismLogLevel logLevel)
+        {
+            DismGeneration dismGeneration = DismUtilities.GetLatestDismGeneration();
```
This probably needs to do nothing if its already been initialized right?

```C#
if(DismGenerationInitialized != DismGeneration.None)
{
    return;
}
```

Or maybe even throw, I'm not sure what `Initialize` will do if you accidentally call it twice...

Shidell commented:
No problem -- everybody is busy. :)

Good call on the Automatic Property. I made that change, and cleaned up the InitializeEx() methods as well, so the overloads cascade instead of repeating the same logic. I also added an additional Exception handler, in the event the user does try to InitializeEx() after already loading a DISM Generation library and before freeing it (Shutdown()).

Any other thoughts or changes you can think of?

josemesona commented:
I've pushed a few commits, double check them and I'll merge!

Shidell commented:
Looks great! I like moving the expanded functionality in DismApiEx, good idea.

josemesona commented:
That's fine, I was just waiting for you to comment.  Thanks for the contribution, I'll push a package.

josemesona commented:
Package will be available at https://www.nuget.org/packages/Microsoft.Dism/1.6.446 in about 10 minutes (it is being "validated" by nuget.org)
